### PR TITLE
Add support for adaptive chunk sizing

### PR DESCRIPTION
**Pull request by erimatnor**
Users can now (optionally) set a target chunk size and TimescaleDB
will try to adapt the interval length of the first open ("time")
dimension in order to reach that target chunk size. If a hypertable
has more than one open dimension, only the first one will have a
dynamically adapting interval.

Users can optionally specify their own function that calculates the
new dimension interval. They can also set a target size of 0 in order
to estimate a suitable target size for a chunk based on available
memory.

### CODE REVIEW

**Comment by andrew-blake** [comment]
Will the estimation of chunk size based upon available memory consider the impact of having more than one hypertable ?

**Comment by erimatnor** [comment]
@andrew-blake We currently do not take into account the number of hypertables in the chunk size estimation. The reason is that there might not be a clear relationship between number of hypertables, concurrent usage, and memory (i.e., not all hypertables might actually have the same query/insert load. Some might even be more or less unused). Instead, we try to make a conservative estimate that would account for some degree of concurrent usage across hypertables. The main goal is for chunks (and their indexes) to fit in memory. It is inevitable that individual chunks (from same or other hypertables) will eventually be swapped out of the shared buffer. Also note that estimation is currently informed by the standard `effective_cache_size` setting, but you can always set a specific "target" chunk size yourself (e.g. `1GB`) that better fits your use case.

That said, we are looking for feedback on adaptive chunking and will probably improve this functionality over next few releases.

**Comment by RobAtticus** [comment]
Won't have much to say about the code, but will run it through benchmarks to see what it comes up with

File: src/chunk_adaptive.c [hunk 1]
```diff
@@ -0,0 +1,551 @@
...
+			 */
+			interval_fillfactor = ((double) max - min) / slice_interval;
+
+			/*
+			 * Extrapolate the size the chunk would have if it spanned the
+			 * entire interval
+			 */
+			extrapolated_chunk_size = chunk_size / interval_fillfactor;
+			size_fillfactor = ((double) extrapolated_chunk_size) / chunk_target_size_bytes;
+
```
Comment (cevian) [inline src/chunk_adaptive.c]: Can we have all these values be print out into the log at DEBUG1 or something. Right now this is totally a black box.

Comment (cevian) [reply src/chunk_adaptive.c]: Maybe even better to have a guc to define whether to output or not.

File: test/expected/chunk_adaptive.out [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+       38 | "_timescaledb_internal"."_hyper_4_38_chunk" | {time,location}      | {"timestamp with time zone",integer} | {NULL,_timescaledb_internal.get_partition_hash} | {"[1485654731569580,1487412902849544)","[-9223372036854775808,1073741823)"} |      409600 |      606208 |             |     1015808
+       39 | "_timescaledb_internal"."_hyper_4_39_chunk" | {time,location}      | {"timestamp with time zone",integer} | {NULL,_timescaledb_internal.get_partition_hash} | {"[1487412902849544,1489171074129508)","[-9223372036854775808,1073741823)"} |      360448 |      573440 |             |      933888
+       40 | "_timescaledb_internal"."_hyper_4_40_chunk" | {time,location}      | {"timestamp with time zone",integer} | {NULL,_timescaledb_internal.get_partition_hash} | {"[1487412902849544,1489171074129508)","[1073741823,9223372036854775807)"}  |      352256 |      548864 |             |      901120
+(20 rows)
+
+SELECT id, hypertable_id, interval_length FROM _timescaledb_catalog.dimension;
```
Comment (cevian) [inline test/expected/chunk_adaptive.out]: Do all these test, check for adapting downward? Should we test adapting upward?

File: src/chunk.c [hunk 3]
```diff
@@ -215,6 +215,52 @@ do_dimension_alignment(ChunkScanCtx *scanctx, Chunk *chunk)
...
+		dim = &hs->dimensions[i];
+
+		if (IS_OPEN_DIMENSION(dim))
+			break;
+
+		dim = NULL;
+	}
+
+	/* Nothing to do if no open dimension */
+	if (NULL == dim)
```
Comment (cevian) [inline src/chunk.c]: Should this be a WARNING? Adaptation on but no open dimensions?

File: src/chunk_adaptive.c [hunk 4]
```diff
@@ -0,0 +1,551 @@
...
+			found = minmax_indexscan(rel, idxrel, attnum, minmax);
+
+		index_close(idxrel, AccessShareLock);
+
+		if (found)
+			break;
+	}
+
+	if (!found)
+		found = minmax_heapscan(rel, atttype, attnum, minmax);
```
Comment (cevian) [inline src/chunk_adaptive.c]: I would not do a heapscan. Too costly. I think the logic should be: Do an indexscan, and if there is no appropriate index, abandon operation with a WARNING in the log.

**Comment by RobAtticus** [comment]
@erimatnor - As I recall from my testing, there was an issue where if the initial chunk size is wildly off from the desired target size, it never really improves (e.g., setting it to 1min when the estimated/target size is several hours). I believe it's because the algorithm ignores chunks that are not at least 1% of the target size. Do you remember the rationale for that threshold?

File: src/chunk_adaptive.c (hunk 4 above)
Comment (erimatnor) [reply src/chunk_adaptive.c]: I assume you mean fall back to current chunk size when you say abandon? We cannot fail the insert.

**Comment by RobAtticus** [comment]
@davidkohn88 @cevian @amytai - This PR is now updated if you all want to take a look. @erimatnor took a cursory look before updating his branch.

One question that is still open is what to do if the table does not have an index to easily find the min/max of the table. Should we attempt to do the minmap heap scan, or just skip the calculation for that chunk (which may actually be all the chunks)?

File: src/chunk.c (hunk 3 above)
Comment (JLockerman) [reply src/chunk.c]: This is deep enough in chunk_create that we probably shouldn't kill the transaction here, I think the actual error belongs in the setting.

File: sql/ddl_api.sql [hunk 5]
```diff
@@ -24,9 +26,19 @@ CREATE OR REPLACE FUNCTION  create_hypertable(
...
+    chunk_target_size       TEXT = NULL,
+    chunk_sizing_func       REGPROC = '_timescaledb_internal.calculate_chunk_interval'::regproc
 ) RETURNS VOID AS '@MODULE_PATHNAME@', 'hypertable_create' LANGUAGE C VOLATILE;
 
+-- Set adaptive chunking. To disable, set chunk_target_size => 'off'.
+CREATE OR REPLACE FUNCTION  set_adaptive_chunking(
+    hypertable                     REGCLASS,
+    chunk_target_size              TEXT,
+    INOUT chunk_sizing_func        REGPROC = '_timescaledb_internal.calculate_chunk_interval'::regproc,
+    OUT chunk_target_size          BIGINT
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply sql/ddl_api.sql]: Does it matter technically? It is just the input `text` converted to bytes so reflects the same information.

**Comment by erimatnor** [comment]
@cevian Fixed most of you comments or clarified. Left some of the stuff the @RobAtticus is responsible for untouched because I think they are non-critical and I am not sure some of them are easily fixed (like combining at least one of the debug outputs).

File: sql/ddl_api.sql (hunk 5 above)
[the remark this answers is not in the document]
Comment (cevian) [reply sql/ddl_api.sql]: It just seems weird (i've never seen a language that allows this) and may break in the future. I'd prefer a different name.

**Comment by svenklemm** [comment]
Why is the calculation based on ```effective_cache_size``` and not ```shared_buffers``` which is the memory postgres actually got assigned?

**Pull request merged**